### PR TITLE
Feat multiple accounts and instances

### PR DESCRIPTION
**Pull request by byawitz**
## What does this PR do?

- Enabling multiple users and Appwrite instances, while keeping comparability with previous `prefs.json` scheme.
- Appwrite cloud is now the default endpoint url, this can be changed by adding the `--self-hosted` flag.
- Easy migrate login credentials with `appwrite login migrate`
- Display all logins using `appwrite login list`
- Change context using `appwrite login change --account-id`

## Test Plan

https://github.com/appwrite/sdk-generator/assets/316103/bf25ba1b-4173-4189-b145-4d35060eef78

### CODE REVIEW

File: templates/cli/lib/commands/generic.js.twig [hunk 1]
```diff
@@ -49,61 +119,74 @@ const whoami = new Command("whoami")
...
+            throw Error('Login ID not found');
         }
 
-        success("Signed in as user with ID: " + account.$id);
+        globalConfig.setCurrentLogin(accountId);
+        success(`Current account is ${accountId}`);
     }));
 
+login
+    .command('migrate')
```
Comment (Meldiron) [inline templates/cli/lib/commands/generic.js.twig:164]: Hmm I dont like migrate command in general. Can we make it automatic? Or have backwards compatibility that combines both keys/files?

Comment (byawitz) [reply templates/cli/lib/commands/generic.js.twig:164]: Hooking migration 
image

**Comment by byawitz** [comment]
Changing multiple account selection to the login command and making it interactive

https://github.com/appwrite/sdk-generator/assets/316103/c5c8f259-5c7a-4059-9cdb-d4aae0929bf8

**Comment by byawitz** [comment]
Interactive multiple account logout

https://github.com/appwrite/sdk-generator/assets/316103/814357d8-3942-420c-905c-02dfad863138

**Comment by Meldiron** [comment]
appwrite login
    If its first time: Normal
    If already has account:
        New or existring

appwrite logout
    - If one, sign out
    - If many, select (indicate current)

File: templates/cli/lib/commands/generic.js.twig [hunk 2]
```diff
@@ -50,77 +134,67 @@ const whoami = new Command("whoami")
 
 const login = new Command("login")
     .description(commandDescriptions['login'])
+    .option(`-sa, --self-hosted`, `Flag for enabling custom endpoint for self hosted instances`)
```
Comment (Meldiron) [inline templates/cli/lib/commands/generic.js.twig:137]:
```suggestion
    .option(`-sh, --self-hosted`, `Flag for enabling custom endpoint for self hosted instances`)
```

File: templates/cli/lib/commands/generic.js.twig [hunk 3]
```diff
@@ -187,21 +263,46 @@ const client = new Command("client")
...
 
+const migrate = async () => {
+    if (!globalConfig.has('endpoint') || !globalConfig.has('cookie')) {
+        return;
+    }
+
+    const endpoint = globalConfig.get('endpoint');
+    const cookie = globalConfig.get('cookie');
+
+    log("Old Appwrite login settings were detected, migrating...");
```
Comment (Meldiron) [inline templates/cli/lib/commands/generic.js.twig:285]: I think we can remove this log

File: templates/cli/lib/commands/generic.js.twig [hunk 4]
```diff
@@ -187,21 +263,46 @@ const client = new Command("client")
...
+        cookie,
+        email: 'legacy'
+    };
+
+    globalConfig.addLogin(id, data);
+    globalConfig.setCurrentLogin(id);
+    globalConfig.delete('endpoint');
+    globalConfig.delete('cookie');
+
+    success(`Account was migrated and it's the current account`);
```
Comment (Meldiron) [inline templates/cli/lib/commands/generic.js.twig:298]: I think we dont need log here

File: templates/cli/lib/questions.js.twig [hunk 5]
```diff
@@ -515,9 +584,35 @@ const questionsMfaChallenge = [
     }
 ];
 
+const questionGetEndpoint = [
+    {
+        type: "input",
+        name: "endpoint",
+        message: "Enter the endpoint of your {{ spec.title|caseUcfirst }} server",
+        default: "http://localhost/v1",
```
Comment (Meldiron) [inline templates/cli/lib/questions.js.twig:592]: We should use cloud as default

Comment (byawitz) [reply templates/cli/lib/questions.js.twig:592]: Cloud is default.

This question is available only when the self-hosted flag is passed. Would it still make sense to change it to Appwrite cloud?

**Comment by byawitz** [comment]
Added headless login with MFA factors

image

File: templates/cli/lib/config.js.twig [hunk 6]
```diff
@@ -408,59 +412,140 @@ class Global extends Config {
...
+        const current = this.getCurrentLogin();
+
+        if (current) {
+            const config = this.get(current);
+
+            return config[key];
+        }
+    }
+
+    setTo(key, value) {
```
Comment (christyjacob4) [inline templates/cli/lib/config.js.twig:540]: what exactly is the need for these new methods? how is it different from the `has` method that already exists?

Comment (byawitz) [reply templates/cli/lib/config.js.twig:540]: This is requires as each one of the multiple account is inside a an object, while the `has` method  will fetch the first level the `hasFrom` will fetch the first-level for a given account, 

### old json
```json
{
  "endpoint": "http://localhost/v1",
  "cookie": "cookie"
}
```


### new json
```json
{
  "6659ece200af3158fff95": {
    "endpoint": "http://cloud.appwrite.io/v1",
    "cookie": "Cookie",
    "email": "user@email.com"
  },
  "6659ece20016e58abe6b": {
    "endpoint": "http://localhost/v1",
    "cookie": "Cookie",
    "email": "legacy"
  },
  "current": "6659ece20016e58abe6b"
}
```

File: templates/cli/lib/sdks.js.twig [hunk 7]
```diff
@@ -99,4 +99,5 @@ const sdkForProject = async () => {
 module.exports = {
   sdkForConsole,
   sdkForProject,
+  questionGetEndpoint,
```
Comment (christyjacob4) [inline templates/cli/lib/sdks.js.twig:102]: Lets move all questions to questions.js file only.

Comment (christyjacob4) [reply templates/cli/lib/sdks.js.twig:102]: Lines 5-28 has declared the array `questionGetEndpoint`

File: templates/cli/lib/questions.js.twig [hunk 8]
```diff
@@ -419,8 +468,96 @@ const questionsLogin = [
...
+                    });
+                }
+            })
+
+            return data.sort((a, b) => Number(b.current) - Number(a.current))
+        },
+        when: (answers) => answers.method === 'select'
+    },
+];
+const questionGetEndpoint = [
```
Comment (christyjacob4) [inline templates/cli/lib/questions.js.twig:501]: it seems we already have `questionGetEndpoint` here. We have to remove it from sdks.js

File: templates/cli/lib/commands/generic.js.twig [hunk 9]
```diff
@@ -122,8 +151,51 @@ const logout = new Command("logout")
...
+        const answers = await inquirer.prompt(questionsLogout);
+
+        if (answers.accounts) {
+            for (let accountId of answers.accounts) {
+                globalConfig.setCurrentLogin(accountId);
+                await singleLogout(accountId);
+            }
+        }
+
+        const leftLogins = globalConfig.getLogins();
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:189]: shouldn't this method be called getSessions() ?

File: templates/cli/lib/commands/generic.js.twig [hunk 10]
```diff
@@ -53,6 +79,8 @@ const loginCommand = async () => {
                 parseOutput: false
             });
         } else {
+            globalConfig.removeLogin(id);
+            globalConfig.setCurrentLogin(oldCurrent);
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:83]:
```suggestion
            globalConfig.setCurrentSession(oldCurrent);
```

File: templates/cli/lib/commands/generic.js.twig (hunk 9 above)
Comment (byawitz) [reply templates/cli/lib/commands/generic.js.twig:189]: This is to set the next account in line as the `current` one

File: templates/cli/lib/commands/generic.js.twig [hunk 11]
```diff
@@ -122,8 +151,51 @@ const logout = new Command("logout")
...
+    })
+    .action(actionRunner(async () => {
+        const logins = globalConfig.getLogins();
+        const current = globalConfig.getCurrentLogin();
+
+        if (current === '') {
+            return;
+        }
+        if (logins.length === 1) {
+            await singleLogout(current);
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:174]: Also singleLogout is not a very good name. Lets just call it `deleteSession` or `logout`

File: templates/cli/lib/config.js.twig [hunk 12]
```diff
@@ -408,59 +415,140 @@ class Global extends Config {
...
+        })
+    }
+
+    addLogin(login, data) {
+        this.set(login, data);
+    }
+
+    removeLogin(login, data) {
+        this.delete(login);
+    }
```
Comment (christyjacob4) [inline templates/cli/lib/config.js.twig:452]: Lets refactor all these method names to use Session / Session 

And deleteSession / logout

File: templates/cli/lib/config.js.twig [hunk 13]
```diff
@@ -399,59 +407,140 @@ class Global extends Config {
...
+            return "";
+        }
+        return this.get(Global.PREFERENCE_CURRENT);
+    }
+
+    setCurrentSession(endpoint) {
+        this.set(Global.PREFERENCE_CURRENT, endpoint);
+    }
+
+    getLoginIds() {
```
Comment (christyjacob4) [inline templates/cli/lib/config.js.twig:421]: shouldn't this be getSessionIds() ?

File: templates/cli/lib/commands/generic.js.twig [hunk 14]
```diff
@@ -122,8 +151,51 @@ const logout = new Command("logout")
...
+        const answers = await inquirer.prompt(questionsLogout);
+
+        if (answers.accounts) {
+            for (let accountId of answers.accounts) {
+                globalConfig.setCurrentSession(accountId);
+                await deleteSession(accountId);
+            }
+        }
+
+        const leftLogins = globalConfig.getSessions();
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:189]:
```suggestion
        const sessions = globalConfig.getSessions();
```

File: templates/cli/lib/commands/generic.js.twig [hunk 15]
```diff
@@ -122,8 +151,51 @@ const logout = new Command("logout")
...
+    globalConfig.removeSession(accountId);
+}
+
+const logout = new Command("logout")
+    .description(commandDescriptions['logout'])
+    .configureHelp({
+        helpWidth: process.stdout.columns || 80
+    })
+    .action(actionRunner(async () => {
+        const logins = globalConfig.getSessions();
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:167]:
```suggestion
        const sessions = globalConfig.getSessions();
```

File: templates/cli/lib/commands/generic.js.twig [hunk 16]
```diff
@@ -122,8 +151,51 @@ const logout = new Command("logout")
...
+        const answers = await inquirer.prompt(questionsLogout);
+
+        if (answers.accounts) {
+            for (let accountId of answers.accounts) {
+                globalConfig.setCurrentSession(accountId);
+                await deleteSession(accountId);
+            }
+        }
+
+        const leftSessions = globalConfig.getSessions();
```
Comment (christyjacob4) [inline templates/cli/lib/commands/generic.js.twig:189]:
```suggestion
        const sessions = globalConfig.getSessions();
```

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**